cpi_pdf_langchain_chatbot.ipynb

```diff
@@ -0,0 +1,555 @@
+       " Document(page_content='The CPI-U increased 3.4% over the 12 months ending April 2024 before seasonal adjustment. The index rose 0.3% in April 2024 on a seasonally adjusted basis—after increasing 0.4% in March.', metadata={'url': 'https://www.investopedia.com/terms/c/consumerpriceindex.asp', 'thumbnail_url': None, 'title': \"Consumer Price Index (CPI): What It Is and How It's Used\", 'description': 'The index rose 0.3% in April on a seasonally adjusted basis, which, was slightly lower than the 0.4% increase in March 2024. The CPI Index is an inflation indicator closely watched by policymakers and financial markets. A related CPI measure is used to calculate cost-of-living adjustments for ...'})]"
+      ]
+     },
+     "execution_count": 28,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# Let's test it out\n",
+    "response = ydc_retriever.invoke(\"British Columbia’s Consumer Price Index (CPI) in April 2024 was 2.9% higher (unadjusted) than in April 2023.  How does this compare to the Canadian CPI?\")\n",
+    "response"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Creating a Vector DB retriever based on data from a PDF File"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We are going to load a PDF file using the PyPDFLoader in Langchain.  We will then use the RecursiveTextSplitter in Langchain to split the documents into chuncks that can be vectorized.  The vectorized chunks of text will be stored in a Facebook AI Similarity Search (FAISS) vector store.  This vector store will be converted into a Langchain retriever."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 31,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from langchain_community.document_loaders import PyPDFLoader\n",
+    "\n",
+    "# The PDF file we are using can be downloaded from: https://www2.gov.bc.ca/assets/gov/data/statistics/economy/cpi/cpi_highlights.pdf\n",
+    "# load the PDF file\n",
+    "loader = PyPDFLoader(\"bc_cpi_highlights.pdf\")\n",
+    "docs = loader.load()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 32,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from langchain_community.vectorstores import FAISS\n",
+    "from langchain_openai import OpenAIEmbeddings\n",
+    "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
+    "\n",
+    "# split the document into chunks, and vectorize these chunks in a FAISS database\n",
```

Comment:
Very nit: but you might consider wrapping these into classes. Maybe towards the end where you can do something like:

```
conversation_rag_chain = PDF_QA_Chain(pdfs=[...], prompt="...",web=YouRetreiver())
conversation_rag_chain.invoque("Question")
```

Reply:
I agree that wrapping everything in a class would be better if this was something that was being used in production, but isn't it better to invoke everything separately for the purposes of a demo?  That way we can explain step-by-step, call the individual retrievers just to demonstrate their functionality, etc, show the output w/ and w/out chat history.  I am open to suggestions here and if you feel that a class is better, I can wrap everything in a class

Reply:
To clarify. i think everything here is great so you can show step by step. My suggestion was that at the end of the notebook you say, okay I'm gonna put this all together in to a class, instantiate it and then use it.

This way, if someone want's to just use the PDF example, they just copy the class, if they want the individual snippets, they are at the beginning

Reply:
I see what you mean.  I misunderstood your initial comment.  Thank you for clarifying.